### PR TITLE
ARROW-12141: [R] Bindings for grepl

### PR DESCRIPTION
**Pull request by ianmcook**
Also `stringr::str_detect()`

### CODE REVIEW

File: r/R/dplyr.R [hunk 1]
```diff
@@ -520,6 +536,60 @@ arrow_stringr_string_replace_function <- function(FUN, max_replacements) {
   }
 }
 
+# this function assigns definitions for the stringr pattern modifier functions
+# (fixed, regex, etc.) plus a required helper function (ensure_opts) in the
+# calling function's environment, where they are required to evaluate `pattern`
```
Comment (ianmcook) [inline r/R/dplyr.R:541]: This is a bit exotic, I know. I can't think of any cleaner or more concise way to assign these helper functions in both of the functions that need them.

**Comment by ianmcook** [comment]
TODO: add tests

File: r/R/dplyr.R (hunk 1 above)
Comment (nealrichardson) [reply r/R/dplyr.R:541]: Why not pull `ensure_opts` out, call it `ensure_stringr_opts <- function(pattern)`, and have your funky eval inside that?

File: r/R/dplyr.R [hunk 2]
```diff
@@ -438,6 +440,48 @@ build_function_list <- function(FUN) {
...
+      ifelse(fixed && !ignore.case, "match_substring", "match_substring_regex"),
+      x,
+      options = list(pattern = pattern)
+    )
+  }
+}
+
+arrow_stringr_string_match_function <- function(FUN) {
+  function(string, pattern, negate = FALSE) {
+    browser()
```
Comment (nealrichardson) [inline r/R/dplyr.R:465]: 🙈

File: r/R/dplyr.R [hunk 3]
```diff
@@ -438,6 +440,48 @@ build_function_list <- function(FUN) {
   )
 }
 
+arrow_r_string_match_function <- function(FUN) {
+  function(pattern, x, ignore.case = FALSE, fixed = FALSE) {
+    if (ignore.case) {
```
Comment (nealrichardson) [inline r/R/dplyr.R:445]: Seems like this logic can also be factored out instead of duplicated.

File: r/R/dplyr.R (hunk 1 above)
Comment (ianmcook) [reply r/R/dplyr.R:541]: D'oh, that seem so obvious now 🤦‍♂️ Thanks

File: r/R/dplyr.R (hunk 2 above)
Comment (ianmcook) [reply r/R/dplyr.R:465]: Hah, yeah very much still `[WIP]`

File: r/R/dplyr.R (hunk 3 above)
Comment (ianmcook) [reply r/R/dplyr.R]: Simplified and factored out in f05321292daa48bd0a58a151819e99ec6a14bc62

File: r/R/dplyr.R (hunk 1 above)
Comment (ianmcook) [reply r/R/dplyr.R]: Done in 5bf0cf4148d2c572b00ca1e998d5a0590d4e5e5d

File: r/tests/testthat/test-dplyr-string-functions.R [hunk 4]
```diff
@@ -100,8 +206,6 @@ test_that("sub and gsub with ignore.case = TRUE and fixed = TRUE", {
 test_that("str_replace and str_replace_all", {
   df <- tibble(x = c("Foo", "bar"))
 
-  library(stringr)
```
Comment (ianmcook) [inline r/tests/testthat/test-dplyr-string-functions.R:103]: This was a dup; it's already loaded at the top of this file

**Comment by ianmcook** [comment]
@github-actions crossbow submit test-r-minimal-build test-r-versions

**Review by nealrichardson — APPROVED** [review]

**Pull request closed without merging**